dvc/remote/base.py

```diff
@@ -541,15 +553,40 @@ def changed_cache(self, checksum):
         return self.changed_cache_file(checksum)
 
     def cache_exists(self, checksums):
-        # NOTE: The reason for such an odd logic is that most of the remotes
-        # take much shorter time to just retrieve everything they have under
-        # a certain prefix(e.g. s3, gs, ssh, hdfs). Other remotes that can
-        # check if particular file exists much quicker, use their own
-        # implementation of cache_exists(see http, local).
-        #
-        # Result of all() might be way too big, so we should walk through
-        # it in one pass.
-        return list(filter(lambda checksum: checksum in checksums, self.all()))
+        """Check if the given checksums are stored in the remote.
+
+        There are two ways of performing this check:
+
+        - Traverse: Get a list of all the files in the remote
+            (traversing the cache directory) and compare it with
+            the given checksums.
+
+        - No traverse: For each given checksum, run the `exists`
+            method and filter the checksums that aren't on the remote.
+            This is done in parallel.
+
+        The reason for such an odd logic is that most of the remotes
+        take much shorter time to just retrieve everything they have under
+        a certain prefix (e.g. s3, gs, ssh, hdfs). Other remotes that can
+        check if particular file exists much quicker, use their own
+        implementation of cache_exists (see http, local).
+
+        Returns:
+            A list with checksums that were found in the remote
+        """
+
+        if self.no_traverse:
+            with ThreadPoolExecutor(max_workers=self.JOBS) as executor:
+                # fmt: off
+                exists = lambda x: self.exists(self.checksum_to_path_info(x))  # noqa: E731,E501
```

Comment:
The reason why this still is going to be slow is because `exists` is establishing connection each time it is called https://github.com/iterative/dvc/blob/master/dvc/remote/ssh/__init__.py#L131. If you take a look at upload/download methods, they accept batches of targets and reuse a single connection for the whole batch. We should do something similar here.